### PR TITLE
Fix for issues #16, #15, #12, #11

### LINKED ISSUE #11: Add toast from bootstrap
For notifications: successful registration and request about changing password

### LINKED ISSUE #12: Fix minor bug with max length of message
Max length of messages is undefined

### LINKED ISSUE #15: Add link to every user reference
Users expect: "If I see a name of user, then I can click it!"
Btw, currently main part of names is unclickable.

### LINKED ISSUE #16: Cannot set user's avatar
There is no way to set user's avatar.

### PR DESCRIPTION
**Pull request by leemurus**
This fixes #16,  fixes #15, fixes #12, fixes #11

### CODE REVIEW

File: app/api/messages.py [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
+        return bad_request('Must include ' + str(required_fields) + ' fields!')
+
     try:
+        # TODO: separate function for validation
         if data['message'].strip() == '':  # Empty message
             return jsonify(False)
 
+        if len(data['message']) > Constants.MAX_MESSAGE_LENGTH:  # Check length
+            return jsonify(False)
+
```
Comment (dmkulazhenko) [inline app/api/messages.py:26]: Exclude validation logic from routes

File: app/api/messages.py [hunk 2]
```diff
@@ -1,23 +1,33 @@
...
         current_user.send_message(
             recipient_room=Room.query.get_or_404(data['room_id']),
             text=data['message']
         )
-    except Exception:
+    except Exception as exception:
```
Comment (dmkulazhenko) [inline app/api/messages.py]: Too broad exception

File: app/api/rooms.py [hunk 3]
```diff
@@ -10,3 +11,12 @@ def get_room_or_create(profile_id):
     profile_user = User.query.get_or_404(profile_id)
     room_id = Room.get_or_create_room(current_user, profile_user)
     return jsonify(room_id)
+
+
+@bp.route('/rooms/get/recipient/<int:room_id>', methods=['GET'])
+def get_recipient_username(room_id):
+    user = Room.query.get(room_id).get_recipient(current_user)
+    if user is None:
+        return bad_request('This room is not chat')
```
Comment (dmkulazhenko) [inline app/api/rooms.py:20]: Maybe 404?

File: app/api/user.py [hunk 4]
```diff
@@ -63,9 +65,23 @@ def update_self_information():
             address=data['address']
         )
 
-        if 'photo' in data['photo']:
-            current_user.set_profile_information(photo=data['photo'])
+    except Exception as exception:
```
Comment (dmkulazhenko) [inline app/api/user.py:68]: Too broad exception

File: app/api/user.py [hunk 5]
```diff
@@ -63,9 +65,23 @@ def update_self_information():
...
+    try:
+        photo = request.files.get('photo')
+        # TODO: separate function for validation
+
+        photo.seek(0, os.SEEK_END)  # Go to the end of file
+        if photo.tell() / 1024 / 1024 > Constants.MAX_PHOTO_SIZE:
+            raise Exception('Max size of photo is {} MB'.format(Constants.MAX_PHOTO_SIZE))
+
+        current_user.set_profile_information(photo=photo)
     except Exception as exception:
```
Comment (dmkulazhenko) [inline app/api/user.py:85]: Too broad exception

File: app/api/user.py [hunk 6]
```diff
@@ -63,9 +65,23 @@ def update_self_information():
...
+@bp.route('/self/update/photo', methods=['POST'])
+@login_required
+def update_photo():
+    try:
+        photo = request.files.get('photo')
+        # TODO: separate function for validation
+
+        photo.seek(0, os.SEEK_END)  # Go to the end of file
+        if photo.tell() / 1024 / 1024 > Constants.MAX_PHOTO_SIZE:
+            raise Exception('Max size of photo is {} MB'.format(Constants.MAX_PHOTO_SIZE))
```
Comment (dmkulazhenko) [inline app/api/user.py:82]: Exclude validation logic from routes

File: app/models.py [hunk 7]
```diff
@@ -110,15 +110,21 @@ def send_message(self, text, recipient_room):
...
-            file_path = os.path.join(Constants.IMAGE_UPLOAD_FOLDER, filename)
-            file_path_db = os.path.join(Constants.IMAGE_DB_FOLDER, filename)
+            # Delete old photo
+            old_path = os.path.join(Constants.IMAGE_UPLOAD_FOLDER, os.path.basename(self.photo))
+            if os.path.isfile(old_path):
+                os.unlink(old_path)
 
+            # Crate folder for photos
             if not os.path.isdir(Constants.IMAGE_UPLOAD_FOLDER):
                 os.makedirs(Constants.IMAGE_UPLOAD_FOLDER)
```
Comment (dmkulazhenko) [inline app/models.py]: Seems to be better to create such folders in create_app()

File: app/models.py [hunk 8]
```diff
@@ -242,15 +248,22 @@ def add_user(self, user):
...
+            # Delete old photo
+            old_path = os.path.join(
+                Constants.ROOM_IMAGE_UPLOAD_FOLDER, os.path.basename(self.photo)
+            )
+            if os.path.isfile(old_path):
+                os.unlink(old_path)
 
+            # Crate folder for photos
             if not os.path.isdir(Constants.ROOM_IMAGE_UPLOAD_FOLDER):
                 os.makedirs(Constants.ROOM_IMAGE_UPLOAD_FOLDER)
```
Comment (dmkulazhenko) [inline app/models.py]: Seems to be better to create such folders in create_app()

File: app/models.py [hunk 9]
```diff
@@ -242,15 +248,22 @@ def add_user(self, user):
...
 
-            photo.data.save(file_path)
+            filename = '{}_{}.{}'.format(self.id, int(time()), photo.filename.rsplit('.', 1)[-1])
+            file_path = os.path.join(Constants.ROOM_IMAGE_UPLOAD_FOLDER, filename)
+            file_path_db = os.path.join(Constants.ROOM_IMAGE_DB_FOLDER, filename)
+
```
Comment (dmkulazhenko) [inline app/models.py:257]: photo.seek(0)?

**Review by dmkulazhenko — CHANGES_REQUESTED** [review]

File: app/api/messages.py (hunk 1 above)
Comment (leemurus) [reply app/api/messages.py:26]: Created new issues #23

File: app/api/user.py (hunk 6 above)
Comment (leemurus) [reply app/api/user.py:82]: Created new issue #23

File: app/models.py (hunk 9 above)
Comment (leemurus) [reply app/models.py:257]: Fixed

File: app/models.py (hunk 7 above)
Comment (leemurus) [reply app/models.py]: Fixed

File: app/models.py (hunk 8 above)
Comment (leemurus) [reply app/models.py]: Fixed

File: app/api/rooms.py (hunk 3 above)
Comment (leemurus) [reply app/api/rooms.py:20]: Don't think so

File: app/api/messages.py (hunk 2 above)
Comment (leemurus) [reply app/api/messages.py]: Reverted to last version. This fix not for this commit.

File: app/api/user.py (hunk 4 above)
Comment (leemurus) [reply app/api/user.py:68]: This fix not for this commit.

File: app/api/user.py (hunk 5 above)
Comment (leemurus) [reply app/api/user.py:85]: This fix not for this commit.

File: app/api/rooms.py (hunk 3 above)
Comment (dmkulazhenko) [reply app/api/rooms.py:20]: Any arguments?

File: app/api/user.py (hunk 4 above)
Comment (dmkulazhenko) [reply app/api/user.py:68]: Create issue?

File: app/api/user.py (hunk 5 above)
Comment (dmkulazhenko) [reply app/api/user.py:85]: Create issue or link commit? or any info?

File: app/api/rooms.py (hunk 3 above)
Comment (leemurus) [reply app/api/rooms.py:20]: It's API request - not user interface. Seems to be better send answer with error message.

File: app/api/user.py (hunk 4 above)
Comment (leemurus) [reply app/api/user.py:68]: Created issue #24

File: app/api/user.py (hunk 5 above)
Comment (leemurus) [reply app/api/user.py:85]: Created issue #24

**Pull request merged**